src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkWasmModule.java

```diff
@@ -0,0 +1,257 @@
+import static java.nio.charset.StandardCharsets.ISO_8859_1;
+
+import com.dylibso.chicory.runtime.ByteArrayMemory;
+import com.dylibso.chicory.runtime.ExportFunction;
+import com.dylibso.chicory.runtime.Instance;
+import com.dylibso.chicory.wasm.ChicoryException;
+import com.dylibso.chicory.wasm.types.MemoryLimits;
+import com.dylibso.chicory.wasm.WasmModule;
+import com.google.common.collect.ImmutableList;
+import com.google.devtools.build.docgen.annot.DocCategory;
+import com.google.devtools.build.lib.concurrent.ThreadSafety.Immutable;
+import java.time.Duration;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import net.starlark.java.annot.StarlarkBuiltin;
+import net.starlark.java.annot.StarlarkMethod;
+import net.starlark.java.eval.EvalException;
+import net.starlark.java.eval.Printer;
+import net.starlark.java.eval.Starlark;
+import net.starlark.java.eval.StarlarkValue;
+
+@Immutable
+@StarlarkBuiltin(
+    name = "wasm_module",
+    category = DocCategory.BUILTIN,
+    doc = "A WebAssembly module loaded by <code>repository_ctx.load_wasm()</code>.")
+final class StarlarkWasmModule implements StarlarkValue {
+  private final StarlarkPath path;
+  private final Object origPath;
+  private final WasmModule wasmModule;
+  private final String allocateFn;
+
+  public StarlarkWasmModule(
+      StarlarkPath path,
+      Object origPath,
+      byte[] moduleContent,
+      String allocateFn)
+      throws EvalException {
+    WasmModule wasmModule;
+    try {
+      wasmModule = com.dylibso.chicory.wasm.Parser.parse(moduleContent);
+    } catch (ChicoryException e) {
+      throw new EvalException(e);
+    }
+    validateModule(wasmModule, allocateFn);
+
+    this.path = path;
+    this.origPath = origPath;
+    this.wasmModule = wasmModule;
+    this.allocateFn = allocateFn;
+  }
+
+  @Override
+  public boolean isImmutable() {
+    return true;
+  }
+
+  @Override
+  public void repr(Printer printer) {
+    printer.append("<wasm_module path=");
+    printer.repr(origPath);
+    printer.append(">");
+  }
+
+  public StarlarkPath getPath() {
+    return path;
+  }
+
+  @StarlarkMethod(
+      name = "path",
+      structField = true,
+      doc = "The path this WebAssembly module was loaded from.")
+  public Object getOrigPath() {
+    return origPath;
+  }
+
+  public StarlarkWasmExecutionResult execute(
+      String functionName,
+      byte[] input,
+      Duration timeout,
+      long memLimitBytes)
+      throws EvalException, InterruptedException {
+    var memLimits = getMemLimits(memLimitBytes);
+    // Perform initialization and execution in a separate thread so it can be interrupted
+    // in case of timeout.
+    var wasmThreadFactory =
+        Thread.ofPlatform().name(Thread.currentThread().getName() + "_wasm").factory();
```

Comment:
FYI: On some platforms (Linux) the thread name is implicitly truncated to 15 characters (16 bytes incl trailing `NUL`).

How would folks feel about changing this to `.name("RepoExecWasm")?
```
$ ps -T -p 2582676 
    PID    SPID TTY          TIME CMD
2582676 2582676 ?        00:00:00 java
2582676 2582677 ?        00:00:00 java
2582676 2582678 ?        00:00:00 GC Thread#0
[...]
2582676 2593224 ?        00:00:00 skyframe-evalua
2582676 2593225 ?        00:00:00 ForkJoinPool-8-
2582676 2593226 ?        00:00:00 ForkJoinPool-8-
2582676 2593227 ?        00:00:04 starlark-reposi
```

Reply:
I won't be truncated in the Starlark profile, which I think is the more common way to inspect Bazel threads. It would be great to be able to link it to the thread that runs the repo rule itself.

Which reminds me that we should also add a profiler span (`Profiler` class) for the WASM execution if you aren't doing that already.

Reply:
OK, added profiling spans for loading/executing wasm.